### PR TITLE
[WIP] feat: add proxy

### LINKED ISSUE #473: Proxy Server
**User story.**

As a client developer integrating with an API which has moved from the mock stage to a prototype or staging instance, I can run Prism in proxy mode, send all that API's traffic through this proxy, which will throw big loud errors if there is an issue with my request, or the response I'm getting back.

As a QA tester managing an e2e suite testing flows between a bunch of APIs, I can funnel all traffic through the Prism proxy, to make sure all the requests and responses coming through the proxy line up with what their API descriptions are advertising, in order to catch contract mismatches quicker, and without relying on those API teams to self report issues.

**Is your feature request related to a problem?**

This is a feature that Prism v1 and v2 had in various forms, and we sadly had to take it out of v3 due to time constraints, and slightly unclear specifications. The two use cases above should help shape thew return of this feature in v3.2 or later. 

**Describe the solution you'd like**

When #429 is done, we should have the means to listen for contract mismatches. Using this mechanism we can have the proxy server listen to requests and using the Prism HTTP Client JS interface, it can make those requests to the real server.

The proxy server should have two modes:

```js
prism proxy {upstream} openapi.yaml --log
prism proxy {upstream} openapi.yaml --error
```

Not sure which should be default right now, but the two modes would work like this:

1. Log would just report mismatches to logs stdout/stderr
2. Error would respond with Prism JSON errors explaining the mismatch

the goal of logs would be to have it running in staging or maybe even production (agh this will need lots of load testing), and error would be more useful for local development and testing environments.

**Additional context**

Blocked by #429 

/cc @ww-daniel-mora for input.

### PR DESCRIPTION
**Pull request by lag-of-death**
Closes #473 

## Checklist

- [ ] Tests have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## What kind of change does this PR introduce?

_Bug fix_, _feature_, or _docs update_

## What is the current behavior? What is the new behavior?

_You can also link to an open issue here for all the info._

## Does this PR introduce a breaking change?

_What changes might users need to make in their application due to this PR?_

### CODE REVIEW

**Comment by lag-of-death** [comment]
@XVincentX, @philsturgeon, I've created this PR so that you can give feedback early :)

File: packages/cli/src/commands/proxy.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import * as signale from 'signale';
+import { CommandModule } from 'yargs';
+import { createMultiProcessPrism, CreatePrismOptions, createSingleProcessPrism } from '../util/createServer';
+import getHttpOperations from '../util/getHttpOperations';
+import options from './options';
+
+const proxyCommand: CommandModule = {
+  describe: 'Start a proxy server with the given spec file',
```
Comment (philsturgeon) [inline packages/cli/src/commands/proxy.ts]: Let's not talk about spec files, are are trying to stick to a unwritten styleguide but we call them "description documents". So "document" is fine.

File: packages/cli/src/commands/proxy.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        type: 'string',
+      })
+      .middleware(async argv => (argv.operations = await getHttpOperations(argv.spec!)))
+      .fail((msg, err) => {
+        if (msg) yargs.showHelp();
+        else signale.fatal(err.message);
+
+        process.exit(1);
+      })
+      .options(options),
```
Comment (philsturgeon) [inline packages/cli/src/commands/proxy.ts]: Sharing all options between mock and proxy is a little confusing as we do not want or need --dynamic. Can we have some shared options and some specific to the command?

**Comment by pytlesk4** [comment]
@philsturgeon is proxy only going to proxy requests? It would be cool if we could proxy/mock requests for a particular spec. It was something we did in the first version of Prism, you could turn mocking on for particular operations. Thought was you could start out mocking your API, and slowly over time turn mocking off for operations and start recieving real data from your API. It was possible in Prism V2, but not as easy as v1. Here is how it was done in classic: https://help.stoplight.io/prism/getting-started/swagger

I don't know if that is out of scope for this PR, or if it is a different mode. But just wanted to get it on your radar.

**Comment by philsturgeon** [comment]
Let's leave this PR comments for comments about the PR. We can chat on #473.

File: packages/cli/src/commands/proxy.ts (hunk 2 above)
Comment (XVincentX) [reply packages/cli/src/commands/proxy.ts]: Yes we can

File: packages/http-server/src/server.ts [hunk 3]
```diff
@@ -4,21 +4,44 @@ import { IHttpOperation } from '@stoplight/types';
 import * as fastify from 'fastify';
 import * as fastifyCors from 'fastify-cors';
 import * as formbodyParser from 'fastify-formbody';
+const proxy = require('fastify-http-proxy');
```
Comment (XVincentX) [inline packages/http-server/src/server.ts]: Use import instead of require

**Comment by lag-of-death** [comment]
alternative approach with `axios`: https://github.com/stoplightio/prism/pull/651

**Comment by XVincentX** [comment]
I'm closing this for now since it does not appear to be a viable alternative.

**Pull request closed without merging**